Index: exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/ItemImpl.java
===================================================================
--- exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/ItemImpl.java	(revision 4647)
+++ exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/ItemImpl.java	(working copy)
@@ -410,6 +410,18 @@
       ConstraintViolationException, RepositoryException
    {
 
+      // Check if checked-in (versionable)
+      if (!parentNode.checkedOut())
+      {
+         throw new VersionException("Node " + parentNode.getPath() + " or its nearest ancestor is checked-in");
+      }
+
+      // Check is locked
+      if (!parentNode.checkLocking())
+      {
+         throw new LockException("Node " + parentNode.getPath() + " is locked ");
+      }
+
       QPath qpath = QPath.makeChildPath(parentNode.getInternalPath(), propertyName);
 
       int state;
@@ -485,18 +497,6 @@
             + locationFactory.createJCRPath(qpath).getAsString(false));
       }
 
-      // Check if checked-in (versionable)
-      if (!parentNode.checkedOut())
-      {
-         throw new VersionException("Node " + parentNode.getPath() + " or its nearest ancestor is checked-in");
-      }
-
-      // Check is locked
-      if (!parentNode.checkLocking())
-      {
-         throw new LockException("Node " + parentNode.getPath() + " is locked ");
-      }
-
       List<ValueData> valueDataList = new ArrayList<ValueData>();
 
       // cast to required type if neccessary
