Index: exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/XASessionImpl.java
===================================================================
--- exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/XASessionImpl.java	(revision 4455)
+++ exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/XASessionImpl.java	(working copy)
@@ -208,9 +208,9 @@
       catch (TransactionException e)
       {
          commitException = e;
-         if (LOG.isDebugEnabled())
+         if (LOG.isErrorEnabled())
          {
-            LOG.debug("Commit Error. Xid:" + xid + ", session: " + getSessionInfo() + ", " + this, e);
+            LOG.error("Commit Error. Xid:" + xid + ", session: " + getSessionInfo() + ", " + this, e);
          }
 
          throw new XASessionException(e.toString(), e.getErrorCode());
