Index: exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/lock/FileSystemLockPersister.java
===================================================================
--- exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/lock/FileSystemLockPersister.java	(revision 3770)
+++ exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/core/lock/FileSystemLockPersister.java	(working copy)
@@ -263,14 +263,7 @@
       {
          log.debug("Stoping FileSystemLockPersister");
       }
-      try
-      {
-         removeAll();
-      }
-      catch (LockException e)
-      {
-         log.error(e.getLocalizedMessage(), e);
-      }
+
       if (log.isDebugEnabled())
       {
          log.debug("FileSystemLockPersister stoped");
