Index: exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/dataflow/session/SessionItemPool.java
===================================================================
--- exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/dataflow/session/SessionItemPool.java	(revision 4711)
+++ exo.jcr.component.core/src/main/java/org/exoplatform/services/jcr/impl/dataflow/session/SessionItemPool.java	(working copy)
@@ -34,6 +34,7 @@
  * @author <a href="mailto:geaz@users.sourceforge.net">Gennady Azarenkov</a>
  * @version $Id: SessionItemPool.java 11907 2008-03-13 15:36:21Z ksm $
  */
+@Deprecated
 final class SessionItemPool
 {
 
@@ -47,7 +48,6 @@
    void remove(String identifier)
    {
       items.remove(identifier);
-      System.gc();
    }
 
    void put(ItemImpl item)
